### PR TITLE
Use System.Text.Json instead of ServiceStack.Text

### PR DESCRIPTION
**Pull request by Ullmie02**
**Changes**
Use `System.Text.Json` everywhere instead of a custom wrapper of `ServiceStack.Text`

### CODE REVIEW

File: Emby.Server.Implementations/Channels/ChannelManager.cs [hunk 1]
```diff
@@ -343,15 +340,16 @@ private List<MediaSourceInfo> GetSavedMediaSources(BaseItem item)
 
             try
             {
-                return _jsonSerializer.DeserializeFromFile<List<MediaSourceInfo>>(path) ?? new List<MediaSourceInfo>();
+                using FileStream jsonStream = File.OpenRead(path);
+                return JsonSerializer.DeserializeAsync<List<MediaSourceInfo>>(jsonStream, JsonDefaults.GetOptions()).GetAwaiter().GetResult();
```
Comment (crobibero) [inline Emby.Server.Implementations/Channels/ChannelManager.cs:344]:
```suggestion
                return JsonSerializer.DeserializeAsync<List<MediaSourceInfo>>(jsonStream, JsonDefaults.GetOptions()).GetAwaiter().GetResult()
                    ?? new List<MediaSourceInfo>();
```

File: Emby.Server.Implementations/Channels/ChannelManager.cs [hunk 2]
```diff
@@ -812,7 +810,8 @@ public async Task<QueryResult<BaseItemDto>> GetChannelItems(InternalItemsQuery q
             {
                 if (_fileSystem.GetLastWriteTimeUtc(cachePath).Add(cacheLength) > DateTime.UtcNow)
                 {
-                    var cachedResult = _jsonSerializer.DeserializeFromFile<ChannelItemResult>(cachePath);
+                    var jsonString = await File.ReadAllTextAsync(cachePath, cancellationToken);
+                    var cachedResult = JsonSerializer.Deserialize<ChannelItemResult>(jsonString);
```
Comment (crobibero) [inline Emby.Server.Implementations/Channels/ChannelManager.cs:814]:
```suggestion
                    await using var fileStream = File.OpenRead(cachePath);
                    var cachedResult = await JsonSerializer.DeserializeAsync<ChannelItemResult>(fileStream, JsonDefaults.GetOptions()).ConfigureAwait(false);
```

File: Emby.Server.Implementations/Channels/ChannelManager.cs [hunk 3]
```diff
@@ -834,7 +833,8 @@ public async Task<QueryResult<BaseItemDto>> GetChannelItems(InternalItemsQuery q
                 {
                     if (_fileSystem.GetLastWriteTimeUtc(cachePath).Add(cacheLength) > DateTime.UtcNow)
                     {
-                        var cachedResult = _jsonSerializer.DeserializeFromFile<ChannelItemResult>(cachePath);
+                        var jsonString = await File.ReadAllTextAsync(cachePath, cancellationToken);
+                        var cachedResult = JsonSerializer.Deserialize<ChannelItemResult>(jsonString);
```
Comment (crobibero) [inline Emby.Server.Implementations/Channels/ChannelManager.cs:837]:
```suggestion
                        await using var fileStream = File.OpenRead(cachePath);
                        var cachedResult = await JsonSerializer.DeserializeAsync<ChannelItemResult>(fileStream, JsonDefaults.GetOptions()).ConfigureAwait(false);
```

File: Emby.Server.Implementations/Library/LiveStreamHelper.cs [hunk 4]
```diff
@@ -83,7 +83,8 @@ public async Task AddMediaInfoWithProbe(MediaSourceInfo mediaSource, bool isAudi
                 if (cacheFilePath != null)
                 {
                     Directory.CreateDirectory(Path.GetDirectoryName(cacheFilePath));
-                    _json.SerializeToFile(mediaInfo, cacheFilePath);
+                    await using FileStream createStream = File.Create(cacheFilePath);
+                    await JsonSerializer.SerializeAsync(createStream, mediaInfo, cancellationToken: cancellationToken).ConfigureAwait(false);
```
Comment (crobibero) [inline Emby.Server.Implementations/Library/LiveStreamHelper.cs:87]:
```suggestion
                    await JsonSerializer.SerializeAsync(createStream, mediaInfo, JsonDefaults.GetOptions(), cancellationToken).ConfigureAwait(false);
```

**Review by crobibero — APPROVED** [review]

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 5]
```diff
@@ -1051,7 +1047,8 @@ public IEnumerable<LocalPlugin> GetLocalPlugins(string path, bool cleanup = true
                     var metafile = Path.Combine(dir, "meta.json");
                     if (File.Exists(metafile))
                     {
-                        var manifest = _jsonSerializer.DeserializeFromFile<PluginManifest>(metafile);
+                        using FileStream jsonStream = File.OpenRead(metafile);
+                        var manifest = JsonSerializer.DeserializeAsync<PluginManifest>(jsonStream, jsonOptions).GetAwaiter().GetResult();
```
Comment (Bond-009) [inline Emby.Server.Implementations/ApplicationHost.cs:1051]: Didn't know there was no sync overload that accepts a `Stream` :thinking:, in this case deserializing from a string/raw bytes is probably better (or at least cleaner)

File: Emby.Server.Implementations/ApplicationHost.cs [hunk 6]
```diff
@@ -101,6 +100,7 @@
 using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Logging;
 using Prometheus.DotNetRuntime;
+using JsonSerializer = System.Text.Json.JsonSerializer;
```
Comment (Bond-009) [inline Emby.Server.Implementations/ApplicationHost.cs:103]: Why are these needed?

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs [hunk 7]
```diff
@@ -335,15 +334,25 @@ private async Task<string> EnsureSeasonInfo(string seriesImdbId, int seasonId, C
...
             return path;
         }
 
+        public static async Task<T> GetDeserializedOmdbResponse<T>(HttpClient httpClient, string url, JsonSerializerOptions jsonOptions, CancellationToken cancellationToken)
+        {
+            using var response = await GetOmdbResponse(httpClient, url, cancellationToken).ConfigureAwait(false);
+            var content = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+
+            // OMDb is sending "N/A" for no empty number fields
+            content = content.Replace("\"N/A\"", "\"\"", StringComparison.InvariantCulture);
```
Comment (Bond-009) [inline MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:352]: Maybe we should make a custom `JsonConverter` for this?
@crobibero

**Review by Bond-009 — CHANGES_REQUESTED** [review]
Maybe we should always put `JsonDefaults.GetOptions()` in a private field for consistency and less function calls? @cvium @crobibero opinions?

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs (hunk 7 above)
Comment (crobibero) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:352]: Almost as amazing as tmdb's api!

Comment (crobibero) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:352]: I have an untested converter, if someone could get a sample response I'll wire it up and add tests

Comment (Ullmie02) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:352]:
```
{"Title":"Chapter 1","Year":"2013","Rated":"TV-MA","Released":"01 Feb 2013","Season":"N/A","Episode":"N/A","Runtime":"55 min","Genre":"Drama","Director":"David Fincher","Writer":"Michael Dobbs (based on the novels by), Andrew Davies (based on the mini-series by), Beau Willimon (created for television by), Beau Willimon, Sam Forman (staff writer)","Actors":"Kevin Spacey, Robin Wright, Kate Mara, Corey Stoll","Plot":"Congressman Francis Underwood has been declined the chair for Secretary of State. He's now gathering his own team to plot his revenge. Zoe Barnes, a reporter for the Washington Herald, will do anything to get her big break.","Language":"English","Country":"USA","Awards":"N/A","Poster":"https://m.media-amazon.com/images/M/MV5BMTY5MTU4NDQzNV5BMl5BanBnXkFtZTgwMzk2ODcxMzE@._V1_SX300.jpg","Ratings":[{"Source":"Internet Movie Database","Value":"8.7/10"}],"Metascore":"N/A","imdbRating":"8.7","imdbVotes":"6736","imdbID":"tt2161930","seriesID":"N/A","Type":"episode","Response":"True"}
```
this should be exactly the one that broke it

File: Emby.Server.Implementations/ApplicationHost.cs (hunk 6 above)
Comment (Ullmie02) [reply Emby.Server.Implementations/ApplicationHost.cs:103]: When I moved this file to the new Serializer, the old one still existed in a namespace which was used in this file.
I'll remove it

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs (hunk 7 above)
Comment (crobibero) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:351]: https://github.com/Ullmie02/jellyfin/pull/1

File: MediaBrowser.Common/Json/Converters/JsonOmdbNotAvailableStringConverter.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            if (reader.TokenType == JsonTokenType.String)
+            {
+                var str = reader.GetString();
+                if (str != null && str.Equals("N/A", StringComparison.OrdinalIgnoreCase))
+                {
+                    return null;
+                }
+            }
+
+            return JsonSerializer.Deserialize<string>(ref reader, options);
```
Comment (Bond-009) [inline MediaBrowser.Common/Json/Converters/JsonOmdbNotAvailableStringConverter.cs:24]: Can't we return `str` inside of the if statement if it isn't `N/A`? @crobibero

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs [hunk 9]
```diff
@@ -459,74 +469,104 @@ private bool IsConfiguredForEnglish(BaseItem item)
...
             public RootObject[] Episodes { get; set; }
 
+            [JsonConverter(typeof(JsonOmdbNotAvailableStringConverter))]
             public string Response { get; set; }
         }
 
-        internal class RootObject
+        public class RootObject
         {
+            [JsonConverter(typeof(JsonOmdbNotAvailableStringConverter))]
```
Comment (Bond-009) [inline MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:494]: Isn't creating a new `JsonSerializerOptions` and adding `JsonOmdbNotAvailableStringConverter` a cleaner solution?

File: tests/Jellyfin.Common.Tests/Jellyfin.Common.Tests.csproj [hunk 10]
```diff
@@ -29,6 +29,7 @@
 
   <ItemGroup>
     <ProjectReference Include="../../MediaBrowser.Common/MediaBrowser.Common.csproj" />
+    <ProjectReference Include="..\..\MediaBrowser.Providers\MediaBrowser.Providers.csproj" />
```
Comment (Bond-009) [inline tests/Jellyfin.Common.Tests/Jellyfin.Common.Tests.csproj:32]:
```suggestion
    <ProjectReference Include="../../MediaBrowser.Providers/MediaBrowser.Providers.csproj" />
```
Consistency

File: tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs [hunk 11]
```diff
@@ -0,0 +1,57 @@
+﻿using System.Text.Json;
+using System.Text.Json.Serialization;
+using MediaBrowser.Common.Json;
+using MediaBrowser.Common.Json.Converters;
+using MediaBrowser.Providers.Plugins.Omdb;
+using Xunit;
+
+namespace Jellyfin.Common.Tests.Json
+{
+    public static class JsonOmdbConverterTests
```
Comment (Bond-009) [inline tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs:10]: I would add at least 1 test where the string isn't `N/A` (and prob also one where the string is `null`)

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.Common/Json/Converters/JsonOmdbNotAvailableStringConverter.cs (hunk 8 above)
Comment (Ullmie02) [reply MediaBrowser.Common/Json/Converters/JsonOmdbNotAvailableStringConverter.cs:26]: Yes we need to return `str`, else we get a stack overflow

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs (hunk 9 above)
Comment (Ullmie02) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs]: Looks definitely cleaner. But maybe we should use the attributes in case this object is deserialized somewhere else and the converter isn't added?

File: Emby.Server.Implementations/LiveTv/EmbyTV/EncodedRecorder.cs [hunk 12]
```diff
@@ -95,7 +93,7 @@ private Task RecordFromFile(MediaSourceInfo mediaSource, string inputFile, strin
             // FFMpeg writes debug/error info to stderr. This is useful when debugging so let's put it in the log directory.
             _logFileStream = new FileStream(logFilePath, FileMode.Create, FileAccess.Write, FileShare.Read, IODefaults.FileStreamBufferSize, true);
 
-            var commandLineLogMessageBytes = Encoding.UTF8.GetBytes(_json.SerializeToString(mediaSource) + Environment.NewLine + Environment.NewLine + commandLineLogMessage + Environment.NewLine + Environment.NewLine);
+            var commandLineLogMessageBytes = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(mediaSource, _jsonOptions) + Environment.NewLine + Environment.NewLine + commandLineLogMessage + Environment.NewLine + Environment.NewLine);
```
Comment (Bond-009) [inline Emby.Server.Implementations/LiveTv/EmbyTV/EncodedRecorder.cs:96]: Can we serialize directly to the `Stream` here?

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs (hunk 7 above)
Comment (crobibero) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:357]: I think this can now be updated to not manually remove `N/A`?

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs (hunk 9 above)
Comment (Bond-009) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs]: These aren't supposed to get use somewhere else. Not sure why they are public now, but they used to be internal

File: Emby.Server.Implementations/LiveTv/EmbyTV/EncodedRecorder.cs [hunk 13]
```diff
@@ -93,8 +93,8 @@ private Task RecordFromFile(MediaSourceInfo mediaSource, string inputFile, strin
             // FFMpeg writes debug/error info to stderr. This is useful when debugging so let's put it in the log directory.
             _logFileStream = new FileStream(logFilePath, FileMode.Create, FileAccess.Write, FileShare.Read, IODefaults.FileStreamBufferSize, true);
 
-            var commandLineLogMessageBytes = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(mediaSource, _jsonOptions) + Environment.NewLine + Environment.NewLine + commandLineLogMessage + Environment.NewLine + Environment.NewLine);
-            _logFileStream.Write(commandLineLogMessageBytes, 0, commandLineLogMessageBytes.Length);
+            await JsonSerializer.SerializeAsync(_logFileStream, mediaSource, _jsonOptions, cancellationToken).ConfigureAwait(false);
+            await _logFileStream.WriteAsync(Encoding.UTF8.GetBytes(Environment.NewLine + Environment.NewLine + commandLineLogMessage + Environment.NewLine + Environment.NewLine), cancellationToken);
```
Comment (Bond-009) [inline Emby.Server.Implementations/LiveTv/EmbyTV/EncodedRecorder.cs:97]:
```suggestion
            await _logFileStream.WriteAsync(Encoding.UTF8.GetBytes(Environment.NewLine + Environment.NewLine + commandLineLogMessage + Environment.NewLine + Environment.NewLine), cancellationToken).ConfigureAwait(false);
```

File: Emby.Server.Implementations/ScheduledTasks/ScheduledTaskWorker.cs [hunk 14]
```diff
@@ -150,7 +144,15 @@ public TaskResult LastExecutionResult
...
                             {
-                                _lastExecutionResult = _jsonSerializer.DeserializeFromFile<TaskResult>(path);
+                                var jsonString = File.ReadAllText(path);
+                                if (!string.IsNullOrWhiteSpace(jsonString))
+                                {
+                                    _lastExecutionResult = JsonSerializer.Deserialize<TaskResult>(jsonString, _jsonOptions);
+                                }
+                                else
+                                {
+                                    _logger.LogDebug("Scheduled Task history file {path} is empty. Skipping deserialization.", path);
```
Comment (Bond-009) [inline Emby.Server.Implementations/ScheduledTasks/ScheduledTaskWorker.cs:154]:
```suggestion
                                    _logger.LogDebug("Scheduled Task history file {Path} is empty. Skipping deserialization.", path);
```

File: Emby.Server.Implementations/ScheduledTasks/ScheduledTaskWorker.cs [hunk 15]
```diff
@@ -573,7 +577,8 @@ private void SaveTriggers(TaskTriggerInfo[] triggers)
 
             Directory.CreateDirectory(Path.GetDirectoryName(path));
 
-            _jsonSerializer.SerializeToFile(triggers, path);
+            using FileStream stream = File.OpenWrite(path);
+            JsonSerializer.SerializeAsync(stream, triggers, _jsonOptions);
```
Comment (Bond-009) [inline Emby.Server.Implementations/ScheduledTasks/ScheduledTaskWorker.cs:581]: ???

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs [hunk 16]
```diff
@@ -335,15 +340,22 @@ private async Task<string> EnsureSeasonInfo(string seriesImdbId, int seasonId, C
...
 
             return path;
         }
 
+        public async Task<T> GetDeserializedOmdbResponse<T>(HttpClient httpClient, string url, CancellationToken cancellationToken)
+        {
+            using var response = await GetOmdbResponse(httpClient, url, cancellationToken).ConfigureAwait(false);
+            var content = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+
+            return JsonSerializer.Deserialize<T>(content, _jsonOptions);
```
Comment (Bond-009) [inline MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:356]: Read as `Stream` and deserialize that (no UTF8->UTF16 convert and possibly less allocations)

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs [hunk 17]
```diff
@@ -459,13 +471,13 @@ private bool IsConfiguredForEnglish(BaseItem item)
             return string.Equals(lang, "en", StringComparison.OrdinalIgnoreCase);
         }
 
-        internal class SeasonRootObject
+        public class SeasonRootObject
```
Comment (Bond-009) [inline MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:474]: Why are these public now?

File: tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        public void Deserialize_Quoted_Number_Success()
+        {
+            const int Number = 8;
+            const string Input = "\"8\"";
+
+            var result = JsonSerializer.Deserialize<int>(Input, _options);
+            Assert.Equal(Number, result);
+        }
+
+        [Fact]
```
Comment (Bond-009) [inline tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs:90]: Opinions on making these `Theory`s instead of `Fact`s?
@crobibero @cvium 
Less descriptive, but also less loc

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs (hunk 18 above)
Comment (crobibero) [reply tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs:90]: IMO it doesn’t matter because the providers already have PRs for removal from server

File: MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs (hunk 17 above)
Comment (crobibero) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:474]: Access inside tests

Comment (Bond-009) [reply MediaBrowser.Providers/Plugins/Omdb/OmdbProvider.cs:474]: Expose internals to the tests (like we do with some other projects)

File: tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs (hunk 18 above)
Comment (Bond-009) [reply tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs:90]: And those tests will move with them, no?

File: Emby.Server.Implementations/ScheduledTasks/ScheduledTaskWorker.cs [hunk 19]
```diff
@@ -573,7 +577,8 @@ private void SaveTriggers(TaskTriggerInfo[] triggers)
 
             Directory.CreateDirectory(Path.GetDirectoryName(path));
 
-            _jsonSerializer.SerializeToFile(triggers, path);
+            var json = JsonSerializer.Serialize(triggers, _jsonOptions);
+            File.WriteAllText(path, json);
```
Comment (BaronGreenback) [inline Emby.Server.Implementations/ScheduledTasks/ScheduledTaskWorker.cs:581]:
```suggestion
            File.WriteAllText(path, json, Encoding.UTF8);
```

**Review by BaronGreenback — APPROVED** [review]

File: tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs (hunk 18 above)
Comment (Ullmie02) [reply tests/Jellyfin.Common.Tests/Json/JsonOmdbConverterTests.cs]: Did move some of the tests to Theorys. Had to be multiple as I couldn't get gernerics to work

File: Emby.Server.Implementations/LiveTv/EmbyTV/ItemDataProvider.cs [hunk 20]
```diff
@@ -61,7 +61,8 @@ private void EnsureLoaded()
         private void SaveList()
         {
             Directory.CreateDirectory(Path.GetDirectoryName(_dataPath));
-            _jsonSerializer.SerializeToFile(_items, _dataPath);
+            using FileStream stream = File.OpenWrite(_dataPath);
+            JsonSerializer.SerializeAsync(stream, _items, _jsonOptions);
```
[the remark this answers is not in the document]
Comment (Ullmie02) [reply Emby.Server.Implementations/LiveTv/EmbyTV/ItemDataProvider.cs]: Can't make the method async because it's used inside a lock statement. Using sync method for file writing now

**Review by Bond-009 — APPROVED** [review]
